In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [2]:
df_scalar_test = pd.read_pickle('../input/nn-to-predict-data-files/df_scalar_test.pkl')
df_scalar_test.head()

/opt/conda/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/opt/conda/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/opt/conda/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/opt/conda/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or 

fc        sd       pso       dso
0    5.787390  0.312003  1.878395 -0.739487
1  172.021181  0.804273 -0.756960  0.212390
2  172.021181  0.804273 -0.756960  0.212390
3    5.787390  0.312003  1.878395 -0.739487
4    2.611647  0.082246  0.600366 -0.682690

In [3]:
df_test_new = pd.read_pickle('../input/molecular-properties-data-create/df_test_new.pkl')
df_test_new.head()

id     molecule_name  atom_index_0  atom_index_1  type      g1_0  \
0  4658147  dsgdb9nsd_000004             2             0  2JHC  0.526855   
1  4658150  dsgdb9nsd_000004             3             0  1JHC  0.526855   
2  4658148  dsgdb9nsd_000004             2             1  1JHC  0.526855   
3  4658151  dsgdb9nsd_000004             3             1  2JHC  0.526855   
4  4658149  dsgdb9nsd_000004             2             3  3JHH  0.526855   

   g2_2_0.01_0  g4_2_0.01_1_1_0  g4_2_0.01_1_8_0  g4_2_0.01_1_16_0  ...  \
0     0.522461              0.0              0.0               0.0  ...   
1     0.522461              0.0              0.0               0.0  ...   
2     0.522461              0.0              0.0               0.0  ...   
3     0.522461              0.0              0.0               0.0  ...   
4     0.522461              0.0              0.0               0.0  ...   

   g4_1_0.2_0.5_1_1  g4_1_0.2_0.5_8_1  g4_1_0.2_0.5_16_1  g2_1.5_0.2_1  \
0               0.0               0.0                0.0      0.928711   
1               0.0               0.0                0.0      0.928711   
2               0.0               0.0                0.0      0.928711   
3               0.0               0.0                0.0      0.928711   
4               0.0               0.0                0.0      0.506836   

   g4_1.5_0.2_1_1_1  g4_1.5_0.2_1_8_1  g4_1.5_0.2_1_16_1  g4_1.5_0.2_0.5_1_1  \
0               0.0               0.0                0.0                 0.0   
1               0.0               0.0                0.0                 0.0   
2               0.0               0.0                0.0                 0.0   
3               0.0               0.0                0.0                 0.0   
4               0.0               0.0                0.0                 0.0   

   g4_1.5_0.2_0.5_8_1  g4_1.5_0.2_0.5_16_1  
0                 0.0                  0.0  
1                 0.0                  0.0  
2                 0.0                  0.0  
3                 0.0                  0.0  
4                 0.0                  0.0  

[5 rows x 217 columns]

In [4]:
cols = df_test_new.columns
df_test_new.drop(columns= cols[5:], inplace=True)
df_test_new.head()

id     molecule_name  atom_index_0  atom_index_1  type
0  4658147  dsgdb9nsd_000004             2             0  2JHC
1  4658150  dsgdb9nsd_000004             3             0  1JHC
2  4658148  dsgdb9nsd_000004             2             1  1JHC
3  4658151  dsgdb9nsd_000004             3             1  2JHC
4  4658149  dsgdb9nsd_000004             2             3  3JHH

In [5]:
df_scalar_test = df_test_new.merge(df_scalar_test, on=df_scalar_test.index)
df_scalar_test.sort_values(by='id')

key_0       id     molecule_name  atom_index_0  atom_index_1  type  \
0              0  4658147  dsgdb9nsd_000004             2             0  2JHC   
2              2  4658148  dsgdb9nsd_000004             2             1  1JHC   
4              4  4658149  dsgdb9nsd_000004             2             3  3JHH   
1              1  4658150  dsgdb9nsd_000004             3             0  1JHC   
3              3  4658151  dsgdb9nsd_000004             3             1  2JHC   
...          ...      ...               ...           ...           ...   ...   
2505526  2505526  7163684  dsgdb9nsd_133885            15             3  2JHC   
2505531  2505531  7163685  dsgdb9nsd_133885            15             4  2JHC   
2505513  2505513  7163686  dsgdb9nsd_133885            15             6  3JHC   
2505520  2505520  7163687  dsgdb9nsd_133885            15             7  2JHC   
2505536  2505536  7163688  dsgdb9nsd_133885            15             8  1JHC   

                 fc        sd       pso       dso  
0          5.787390  0.312003  1.878395 -0.739487  
2        172.021181  0.804273 -0.756960  0.212390  
4          2.611647  0.082246  0.600366 -0.682690  
1        172.021181  0.804273 -0.756960  0.212390  
3          5.787390  0.312003  1.878395 -0.739487  
...             ...       ...       ...       ...  
2505526    1.598614  0.080681  0.137412 -0.124142  
2505531    1.610809  0.081613  0.136313 -0.125660  
2505513    3.605474  0.004166  0.189682 -0.237708  
2505520    2.023700  0.082079  0.088202 -0.079010  
2505536  104.511036  0.300841  0.330784  1.105895  

[2505542 rows x 10 columns]

In [6]:
df_scalar_test.drop(columns= ['key_0', 'id'], inplace = True)

In [7]:
df_scalar_test.head()

molecule_name  atom_index_0  atom_index_1  type          fc        sd  \
0  dsgdb9nsd_000004             2             0  2JHC    5.787390  0.312003   
1  dsgdb9nsd_000004             3             0  1JHC  172.021181  0.804273   
2  dsgdb9nsd_000004             2             1  1JHC  172.021181  0.804273   
3  dsgdb9nsd_000004             3             1  2JHC    5.787390  0.312003   
4  dsgdb9nsd_000004             2             3  3JHH    2.611647  0.082246   

        pso       dso  
0  1.878395 -0.739487  
1 -0.756960  0.212390  
2 -0.756960  0.212390  
3  1.878395 -0.739487  
4  0.600366 -0.682690

In [8]:
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

In [9]:
df_scalar_test = reduce_mem_usage(df_scalar_test)

Mem. usage decreased to 81.24 Mb (41.4% reduction)


In [10]:
mulliken_test = pd.read_pickle('../input/making-mulliken-charges-dataset/mulliken_test.pkl')
mulliken_test.head()

molecule_name  atom_index  mulliken_charge
12  dsgdb9nsd_000004           0        -0.420117
13  dsgdb9nsd_000004           1        -0.420117
14  dsgdb9nsd_000004           2         0.200479
15  dsgdb9nsd_000004           3         0.200479
79  dsgdb9nsd_000015           0        -0.224901

In [11]:
mulliken_test = reduce_mem_usage(mulliken_test)

Mem. usage decreased to 14.95 Mb (24.0% reduction)


In [12]:
data_folder = '../input/champs-scalar-coupling'
mulliken_train = pd.read_csv(f'{data_folder}/mulliken_charges.csv')
train = pd.read_csv(f'{data_folder}/train.csv')
df_scalar_train = pd.read_csv(f'{data_folder}/scalar_coupling_contributions.csv')

# train.head()   # has the y-data (ie, scalar_coupling_constant)

In [13]:
mulliken_train = reduce_mem_usage(mulliken_train)
train = reduce_mem_usage(train)
df_scalar_train = reduce_mem_usage(df_scalar_train)

Mem. usage decreased to 16.09 Mb (54.2% reduction)
Mem. usage decreased to 106.62 Mb (50.0% reduction)
Mem. usage decreased to 115.50 Mb (59.4% reduction)


Merging the Mulliken charges to train and test

In [14]:
df_scalar_test = df_scalar_test.merge(mulliken_test, left_on= ['molecule_name', 'atom_index_0'], right_on= ['molecule_name', 'atom_index'])
df_scalar_test = df_scalar_test.merge(mulliken_test, left_on= ['molecule_name', 'atom_index_1'], right_on= ['molecule_name', 'atom_index'])
# df_scalar_test.head()

In [15]:
df_scalar_train = df_scalar_train.merge(mulliken_train, left_on= ['molecule_name', 'atom_index_0'], right_on= ['molecule_name', 'atom_index'])
df_scalar_train = df_scalar_train.merge(mulliken_train, left_on= ['molecule_name', 'atom_index_1'], right_on= ['molecule_name', 'atom_index'])
# df_scalar_train.head()

In [16]:
df_scalar_test.drop(columns= ['molecule_name', 'atom_index_0', 'atom_index_1', 'atom_index_x', 'atom_index_y'], inplace=True)
df_scalar_train.drop(columns= ['molecule_name', 'atom_index_0', 'atom_index_1', 'atom_index_x', 'atom_index_y'], inplace=True)
# df_scalar_train.head()

In [17]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
le.fit(df_scalar_train['type'])
le.classes_

array(['1JHC', '1JHN', '2JHC', '2JHH', '2JHN', '3JHC', '3JHH', '3JHN'],
      dtype=object)

In [18]:
df_scalar_test['type'] = le.transform(df_scalar_test['type']) 
df_scalar_train['type'] = le.transform(df_scalar_train['type']) 
df_scalar_train.head()

type        fc        sd       pso       dso  mulliken_charge_x  \
0     0  83.00000  0.254639  1.258789  0.271973           0.133911   
1     0  83.00000  0.254639  1.258789  0.271973           0.133911   
2     0  83.00000  0.254639  1.258789  0.271973           0.133911   
3     0  83.00000  0.254639  1.258789  0.271973           0.133911   
4     3 -11.03125  0.353027  2.857422 -3.433594           0.133911   

   mulliken_charge_y  
0          -0.535645  
1          -0.535645  
2          -0.535645  
3          -0.535645  
4           0.133911

Fitting a Neural Network to the above train data

In [19]:
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import cross_val_score
from keras.callbacks import EarlyStopping, ReduceLROnPlateau
from keras.callbacks import ModelCheckpoint
from keras.models import load_model
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, LeakyReLU
from keras.layers import Dropout
from keras.callbacks import EarlyStopping
from keras.callbacks import ModelCheckpoint
import matplotlib.pyplot as plt
from keras.models import load_model

Using TensorFlow backend.


In [ ]:
import keras.backend as K

def custom_metric(y_true, y_pred):
    avg_loss = K.mean(K.abs(y_pred-y_true))
    return K.log(avg_loss)

In [20]:
def model_(m):
    model = Sequential()

    model.add(Dense(50, input_dim=m))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(100))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(250))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(500))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(1000))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(500))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(250))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    
    model.add(Dense(100))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    model.add(Dense(50))
    model.add(BatchNormalization())
    model.add(LeakyReLU(0.1))
    
    model.add(Dense(1, activation='linear'))
    model.compile(loss='mse', optimizer='adam', metrics = [custom_metric])
    return model

In [21]:
model = model_(df_scalar_train.shape[1])

In [22]:
n_epochs = 10
batch_size = 2000
es = EarlyStopping(monitor='val_loss', verbose=1, patience=5)  # , mode='min'
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=10, min_lr=1e-6)
mc = ModelCheckpoint('best_model.h5', monitor='val_loss', verbose=0, save_best_only=True)  # , mode='min'
# fit model
history = model.fit(df_scalar_train, train['scalar_coupling_constant'], validation_split=0.2, epochs=n_epochs, shuffle=True,
                    batch_size = batch_size, verbose=0, callbacks=[es, reduce_lr, mc])

In [24]:
best_model = load_model('best_model.h5')

In [28]:
op_metrics = best_model.evaluate(df_scalar_train, train['scalar_coupling_constant'])

op_metrics

[0.13533528323661, -1.241083654079616]

From the above metrics, it's found that the:
- average mse on prediction is: 0.1353

  
- average log loss is: *-1.2411*

  This log loss is defined by the custom_metric function above

## Metrics Interpretation & Further Improvements

The regressed parameters given by the model are a good fit compared to a simple model whose log loss would be very high (1 or above).

However, looking at the competition results where the Leaderboard touched log-loss in the range of [-3, -3.5], there is certainly a room for improvement such as:

1. In our model, we only used Mulliken charges fitted on test. Similar model fitting of the dipole moment data can be done so that we can even use it as a parameter.
2. Improving the model architecture can be done. Presently, we are only creating an MLP network. However, using custom network (using PyTorch) can help in achieving the required precision in our Scalar Coupling predictions.